File: src/harpy/ast/statements/if_statement.py
from typing import override

from ..expressions import Expression
from .statement import Statement


class IfStatement(Statement):
    """Represents an `if` statement with optional `elseif` and `else` conditions."""

    _ifcond: Expression
    _elifs: list[tuple[Expression, list[Statement]]]
    _ifbody: list[Statement]
    _elsebody: list[Statement]

    def __init__(
        self,
        ifcond: Expression,
        elifs: list[tuple[Expression, list[Statement]]],
        ifbody: list[Statement],
        elsebody: list[Statement],
    ):
        self._ifcond = ifcond
        self._elifs = elifs
        self._ifbody = ifbody
        self._elsebody = elsebody

    @override
    def print(self) -> str:
        outer_stmt = f"if {self._ifcond.print()}\n"
        for stmt in self._ifbody:
            outer_stmt += stmt.print() + "\n"
        for cond, body in self._elifs:
            elifbody = "\n".join([stmt.print() for stmt in body])
            outer_stmt += f"elseif {cond.print()}\n{elifbody}\n"
        if len(self._elsebody) > 0:
            outer_stmt += "else\n"
            for stmt in self._elsebody:
                outer_stmt += f"{stmt.print()}\n"

        return outer_stmt + "endif"

File: Harpy/AST/Statements/IfStatement.cs
using Harpy.AST.Expressions;
using Harpy.CodeGen;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Harpy.AST.Statements;

public class IfStatement : Statement
{
    private readonly List<Statement> _elseBody;
    private readonly List<Tuple<Expression, List<Statement>>> _elseIfConditions;
    private readonly List<Statement> _ifBody;
    private readonly Expression _ifCondition;

    public IfStatement(Expression ifCondition,
        List<Tuple<Expression, List<Statement>>> elseIfConditions,
        List<Statement> ifBody,
        List<Statement> elseBody) : base([])
    {
        _ifCondition = ifCondition;
        _elseIfConditions = elseIfConditions;
        _ifBody = ifBody;
        _elseBody = elseBody;

        _ifCondition.Parent = this;
        Children.Add(_ifCondition);

        foreach (var statement in _ifBody)
        {
            statement.Parent = this;
            Children.Add(statement);
        }

        foreach (var (condition, statements) in _elseIfConditions)
        {
            condition.Parent = this;
            Children.Add(condition);

            foreach (var statement in statements)
            {
                statement.Parent = this;
                Children.Add(statement);
            }
        }

        foreach (var statement in _elseBody)
        {
            statement.Parent = this;
            Children.Add(statement);
        }
    }

    public override string PrettyPrint(int indent = 0)
    {
        var result = NodeLine(indent) + "IfStatement(\n";

        result += BlankLine(indent + 1) + "ifCondition\n" + ChildNodeLine(indent + 1) +
                  _ifCondition.PrettyPrint(indent + 2) + "\n";

        if (_ifBody.Count > 0)
        {
            result += BlankLine(indent + 1) + "ifBody\n";
            result = _ifBody.Aggregate(result,
                (current, stmt) => current + ChildNodeLine(indent + 1) + stmt.PrettyPrint(indent + 2) + "\n");
        }

        if (_elseIfConditions.Count > 0)
        {
            result += BlankLine(indent + 1) + "elseIfConditions\n";
            foreach (var (condition, statements) in _elseIfConditions)
            {
                result += BlankLine(indent + 2) + "condition\n" + ChildNodeLine(indent + 2) +
                          condition.PrettyPrint(indent + 3) + "\n";
                if (statements.Count <= 0) continue;
                result += BlankLine(indent + 2) + "body\n";
                result = statements.Aggregate(result,
                    (current, stmt) => current + ChildNodeLine(indent + 2) + stmt.PrettyPrint(indent + 3) + "\n");
            }
        }

        if (_elseBody.Count > 0)
        {
            result += BlankLine(indent + 1) + "elseBody\n";
            result = _elseBody.Aggregate(result,
                (current, stmt) => current + ChildNodeLine(indent + 1) + stmt.PrettyPrint(indent + 2) + "\n");
        }

        result += BlankLine(indent) + ")";
        return result;
    }

    public override StatementSyntax WalkStatement(CodeGenContext context)
    {
        // TODO: Implement if statement code generation
        throw new NotImplementedException("IfStatement.WalkStatement not yet implemented");
    }
}